### PR TITLE
[B+C] Add Ban Reason to "Vanilla" Bukkit - Fixes BUKKIT-4285

### PR DESCRIPTION
**Pull request by theminecoder**
**The Issue:**
Currently, there is no way to set the ban reason and banning player without using an external plugin.

**Justification:**
By having 1 less plugin to install this will increase (not by much) the server speed of many servers as they will no longer have to write to the databases that they use to store their own bans (which if they are using external servers, can become very laggy for muiltiple write requests)

**Breakdown:**
[C]This just added the required methods to keep current system working as well as new bans.
[B]This added the required changes to the ban command to parse reason and banning player.
[JD]This addresses changes to official javadoc for OfflinePlayer.setBanned - Closed

**Testing:**
Compiled server and started banning users with random reasons. Checked banned-players.txt for verification.
Also banned some with no reason to test if both options work.

**Relevant PR(s):**
https://github.com/Bukkit/CraftBukkit/pull/1250
https://github.com/Bukkit/Bukkit/pull/935
https://github.com/Bukkit/Bukkit-JavaDoc/pull/43 - Closed

**JIRA Ticket:**
https://bukkit.atlassian.net/browse/BUKKIT-4813 - Closed
https://bukkit.atlassian.net/browse/BUKKIT-4285 - Not mine but suits

### CODE REVIEW

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 1]
```diff
@@ -34,6 +34,23 @@
...
+    /**
+     * Bans or unbans this player
+     *
+     * @param banned true if banned
+     * @param banner who banned them
+     */
+    public void setBanned(boolean banned, String banner);
+    
+    /**
+     * Bans or unbans this player
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: Both this and the above added method should have JavaDocs detailing what happens to the "banner" and "reason" arguments in the event of "banned" being false. Maybe a link to the setBanned() method would be idea as well.

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 2]
```diff
@@ -27,8 +28,18 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder();
+            for(String arg: args) {
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Smaller diff (slightly) to do this: (needs reformatting, typed free-hand)

``` java
for(int i=1;i<args.length;i++){
    reason.append(args[i]).append(" ");
}
```

Edit: added a missing `.append(" ")`

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 3]
```diff
@@ -27,8 +28,18 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
+                if(!arg.equals(args[0])) {
+                    reason.append(arg + " ");
+                }
+            }
+            reason.replace(reason.lastIndexOf(" "), reason.lastIndexOf(" "), "");
+            Bukkit.getOfflinePlayer(args[0]).setBanned(true, sender.getName(), reason.toString());
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: You could take out the `replace` call and make this one line... if you wanted:

``` java
Bukkit.getOfflinePlayer(args[0]).setBanned(true, sender.getName(), reason.substring(0, reason.length()-1));
```

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 1 above)
Comment (theminecoder) [inline src/main/java/org/bukkit/OfflinePlayer.java]: I would love to....but I cannot remember for the life of me how to do it. Tips would be great.

**Comment by theminecoder** [comment]
added the code changes...gimme a sec to test them

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 4]
```diff
@@ -34,6 +34,23 @@
...
     public void setBanned(boolean banned);
+    
+    /**
+     * Bans or unbans this player
+     *
+     * @param banned true if banned
+     * @param banner who banned them
+     */
+    public void setBanned(boolean banned, String banner);
+
```
Comment (Wolvereness) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ``` ```
```

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 5]
```diff
@@ -34,6 +34,23 @@
      * @param banned true if banned
      */
     public void setBanned(boolean banned);
+
```
Comment (Wolvereness) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ``` ```
```

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 1 above)
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: In eclipse it (should) be as simple as typing `#` then `setBanned` and selecting the correct method signature. 

If that fails: `{@link #setBanned(boolean)}` should work

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 5 above)
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: ^ Means extra lines

Comment (theminecoder) [inline src/main/java/org/bukkit/OfflinePlayer.java]: I feel like your having issues commenting.

On 1 Oct 2013, at 3:13 pm, Wesley Wolfe notifications@github.com wrote:

In src/main/java/org/bukkit/OfflinePlayer.java:

> @@ -34,6 +34,23 @@
>       \* @param banned true if banned
>       */
>      public void setBanned(boolean banned);
> +

 —
Reply to this email directly or view it on
GitHubhttps://github.com/Bukkit/Bukkit/pull/935/files#r6678437
.

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 1 above)
Comment (theminecoder) [inline src/main/java/org/bukkit/OfflinePlayer.java]: But in where?

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 6]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder();
+            for(int i=1; i<args.length; i++) {
+                reason.append(args[i]).append(" ");
+            }
+            Bukkit.getOfflinePlayer(args[0]).setBanned(true, sender.getName(), reason.substring(0, reason.length()-1));
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Needs reformatting with spaces...

I intentionally didn't make sure it was to spec when I typed it free-hand.

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 7]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder();
+            for(int i=1; i<args.length; i++) {
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: https://github.com/Bukkit/Bukkit/pull/935/files#r6678458

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 1 above)
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: The JavaDocs... You need an accurate description as to what happens to the other variables if "banned==false", and a link to the `setBanned(boolean)` method alongside something that says (something along the lines of) "For unbanning it is recommended to use <insert link here>"

Yes, there are a lot of somethings.

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 5 above)
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: If you look at it on GitHub you'll see he's adding some "blank" comments where you have extra spaces.

**Comment by theminecoder** [comment]
there, that should do it

so after a day of testing it seems to be working great... should be ready to close :D..its up to you now @turt2live

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 8]
```diff
@@ -29,13 +29,30 @@
     public boolean isBanned();
 
     /**
-     * Bans or unbans this player
+     * Bans or unbans this player using default settings. For more settings use {@link #setBanned(boolean, String, String)}. Best for unbanning.
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: Existing javadocs should not be changed (in general) in a PR, that's what Bukkit-Javadoc is for.

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 9]
```diff
@@ -29,13 +29,30 @@
...
      *
      * @param banned true if banned
      */
     public void setBanned(boolean banned);
 
     /**
+     * Bans or unbans this player using default reason. For more settings use {@link #setBanned(boolean, String, String)}. Banner will only be used if banning.
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: I would reword this to say something like "to set a ban reason use <method>". Also 78 column limit.

File: src/main/java/org/bukkit/OfflinePlayer.java [hunk 10]
```diff
@@ -29,13 +29,30 @@
...
+     *
+     * @param banned true if banned
+     * @param banner who banned them
+     */
+    public void setBanned(boolean banned, String banner);
+
+    /**
+     * Bans or unbans this player. Banner and reason will only be used if banning.
```
Comment (turt2live) [inline src/main/java/org/bukkit/OfflinePlayer.java]: `if passing true for banned`

Slightly better wording. Also 78 column limit.

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 11]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder();
+            for(int i=1;i<args.length;i++) {
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: Formatting.

`for (int i = 1; i < args.length; i++;) {`

**Comment by theminecoder** [comment]
all done @turt2live
- Sent from Eclipse via Mylyn

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 12]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder();
```
Comment (ST-DDT) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: How about doing it this way:

```
        if (args.length > 1){
            StringBuilder reason = new StringBuilder(args[1]); 
            for(int i = 2; i < args.length ; i++) {
                reason.append(" ").append(args[i]);
            }
            Bukkit.getOfflinePlayer(args[0]).setBanned(true,sender.getName(),reason.toString());
        }
```

File: src/main/java/org/bukkit/OfflinePlayer.java (hunk 8 above)
Comment (Wolvereness) [inline src/main/java/org/bukkit/OfflinePlayer.java]: @turt2live is wrong here. JavaDocs that directly relate to a change in functionality belong in the PR that changes the functionality. Adding a `@see` tag would be appropriate to put into this PR.

The exception would only be (sometimes) when a PR is only on CB side (like javadocing that an event will now fire in an additional circumstance).

File: src/main/java/org/bukkit/command/defaults/BanCommand.java (hunk 12 above)
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java]: ^ This

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 13]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder(args[1]); 
+            for(int i = 2; i < args.length ; i++) {
+                reason.append(" ").append(args[i]);
+            }
+            Bukkit.getOfflinePlayer(args[0]).setBanned(true,sender.getName(),reason.toString());
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java:36]: Spaces after commas please

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 14]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
+        } else {
+            Bukkit.getOfflinePlayer(args[0]).setBanned(true,sender.getName());
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java:38]: Spaces after commas

File: src/main/java/org/bukkit/command/defaults/BanCommand.java [hunk 15]
```diff
@@ -27,8 +28,15 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return false;
         }
 
-        // TODO: Ban Reason support
-        Bukkit.getOfflinePlayer(args[0]).setBanned(true);
+        if (args.length > 1){
+            StringBuilder reason = new StringBuilder(args[1]); 
+            for(int i = 2; i < args.length ; i++) {
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/BanCommand.java:33]: `args.length ;` should not have a space before the semicolon

**Comment by turt2live** [comment]
What would be appreciated is a way to get the ban information on a player (the BanEntry class).

**Comment by turt2live** [comment]
We've decided to pull this a different way, but we appreciate the work put into this!

Thanks!

https://github.com/Bukkit/Bukkit/commit/75427e084d0f2a444375324b0ef7334aa66d93be
